README.md
# marketmood_api_client2283

"MarketMood API Client" is a Python package that provides a complete set of tools for analysing market sentiment based on news data. The package will allow users to access market sentiment data from the latest hundred news stories of any company they are interested in. Also, users can use this package to see the correlation between the sentiment score and the company's stock price.


## Installation

```bash
$ pip install marketmood_api_client2283
```

## Usage

- TODO

## Contributing

Interested in contributing? Check out the contributing guidelines. Please note that this project is released with a Code of Conduct. By contributing to this project, you agree to abide by its terms.

## License

`marketmood_api_client2283` was created by Mengyan Xu. It is licensed under the terms of the MIT license.

## Credits

`marketmood_api_client2283` was created with [`cookiecutter`](https://cookiecutter.readthedocs.io/en/latest/) and the `py-pkgs-cookiecutter` [template](https://github.com/py-pkgs/py-pkgs-cookiecutter).


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.1.0 (12/12/2023)

- First release of `marketmood_api_client2283`!

tests/marketmood_api_client2283.py
import pandas as pd
import os
import requests
import json
from textblob import TextBlob
from datetime import datetime
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np


def _fetch_financial_sentiment_data(company_name):
    """
    Fetches financial market data for a specific company and analyzes sentiment of the articles using VADER.
    Retrieves the last 100 news items about the company, including the news time and URL.
    :param company_name: Name of the company.
    :return: DataFrame with article details, sentiment scores, publication times, and URLs.
    """
    api_endpoint = 'https://newsapi.org/v2/everything'
    api_key = "f1d6a48ce4574b4ba30eac6e43162957"  
    params = {
        'q': company_name,
        'apiKey': api_key,
        'language': 'en',
        'sortBy': 'publishedAt',
        'pageSize': 100  # Fetch up to 100 articles
    }

    try:
        response = requests.get(api_endpoint, params=params)
        response.raise_for_status()
    except requests.exceptions.HTTPError as http_err:
        print(f'HTTP error occurred: {http_err}')
        return pd.DataFrame()
    except Exception as err:
        print(f'Other error occurred: {err}')
        return pd.DataFrame()

    articles = response.json().get('articles', [])
    data = []
    analyzer = SentimentIntensityAnalyzer()

    for article in articles:
        title = article.get('title', '') or ''
        description = article.get('description', '') or ''
        published_at = datetime.strptime(article.get('publishedAt', '')[:10], '%Y-%m-%d').date() if article.get('publishedAt') else None
        url = article.get('url', '')

        text = f"{title} {description}".strip()
        sentiment_score = analyzer.polarity_scores(text)['compound']
        sentiment_category = 'Positive' if sentiment_score > 0.05 else 'Negative' if sentiment_score < -0.05 else 'Neutral'

        data.append({
            'title': title,
            'description': description,
            'published_at': published_at,
            'sentiment_score': sentiment_score,
            'sentiment_category': sentiment_category,
            'url': url
        })

    df = pd.DataFrame(data)
    return df


# Example usage
company_name = "Tesla"
df = _fetch_financial_sentiment_data(company_name)
df.head()


def classify_average_sentiment(df):
    """
    Computes the average sentiment score and classifies it into levels.
    :param df: DataFrame with sentiment scores.
    :return: A tuple containing the average sentiment score and its classification.
    """
    if df.empty:
        return (None, "No Data")

    average_score = df['sentiment_score'].mean()

    if average_score <= -0.5:
        sentiment_level = "Strongly Negative"
    elif -0.5 < average_score <= -0.1:
        sentiment_level = "Negative"
    elif -0.1 < average_score < 0.1:
        sentiment_level = "Neutral"
    elif 0.1 <= average_score < 0.5:
        sentiment_level = "Positive"
    else:
        sentiment_level = "Strongly Positive"

    return (average_score, sentiment_level)


# +
# Fetch sentiment data
company_name = "Apple"
sentiment_data = _fetch_financial_sentiment_data(company_name)

# Classify average sentiment
average_sentiment, sentiment_level = classify_average_sentiment(sentiment_data)

print(f"Average Sentiment Score for {company_name}: {average_sentiment}")
print(f"Sentiment Level: {sentiment_level}")

# +
# Get the company's stock price and see the correlation between the sentiment scores and the stock price

# +
import yfinance as yf

def fetch_stock_prices(symbol, period='1mo'):
    stock = yf.Ticker(symbol)
    df = stock.history(period=period)
    df.reset_index(inplace=True)
    return df[['Date', 'Close']]


# -

def align_data(df_sentiment, df_stock_prices):
    df_sentiment['date'] = pd.to_datetime(df_sentiment['published_at']).dt.date
    df_stock_prices['date'] = pd.to_datetime(df_stock_prices['Date']).dt.date
    df_combined = pd.merge(df_sentiment, df_stock_prices, left_on='date', right_on='date')
    return df_combined


def analyze_correlation(df_combined):
    return df_combined[['sentiment_score', 'Close']].corr()


# +
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

def plot_data(df_combined):
    plt.figure(figsize=(12, 6))

    # Ensure data is in correct format for Matplotlib
    dates = df_combined['date'].values
    sentiment_scores = df_combined['sentiment_score'].values
    stock_prices = df_combined['Close'].values

    # Create a plot with two y-axes
    fig, ax1 = plt.subplots()

    color = 'tab:red'
    ax1.set_xlabel('Date')
    ax1.set_ylabel('Sentiment Score', color=color)
    ax1.plot(dates, sentiment_scores, color=color, marker='o')
    ax1.tick_params(axis='y', labelcolor=color)

    ax2 = ax1.twinx()  # Instantiate a second axes that shares the same x-axis
    color = 'tab:blue'
    ax2.set_ylabel('Stock Price', color=color)
    ax2.plot(dates, stock_prices, color=color, marker='o')
    ax2.tick_params(axis='y', labelcolor=color)

    plt.title('Sentiment Score and Stock Price Over Time')
    fig.tight_layout()  
    plt.show()



# -

# # Example Usage

# +
# Example usage
company_name = "TSLA"  # Tesla
stock_symbol = "TSLA"

# Fetch Data
df_stock_prices = fetch_stock_prices(stock_symbol)
df_sentiment = _fetch_financial_sentiment_data(company_name) 

# Align and Combine Data
df_combined = align_data(df_sentiment, df_stock_prices)

# Analyze Correlation
correlation = analyze_correlation(df_combined)
print(correlation)

# Plot Data 
plot_data(df_combined)

# -








tests/test_marketmood_api_client2283.py
import pytest
import pandas as pd
import os
import requests
import json
from textblob import TextBlob
from datetime import datetime
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer

from marketmood_api_client2283 import classify_average_sentiment
from marketmood_api_client2283 import _fetch_financial_sentiment_data
from marketmood_api_client2283 import fetch_stock_prices  


# +
import pytest
import pandas as pd

def test_fetch_financial_sentiment_data_simple():
    company_name = "Tesla"
    
    # Call the function
    df = _fetch_financial_sentiment_data(company_name)

    # Check if the result is a DataFrame
    assert isinstance(df, pd.DataFrame)

    # Check if the DataFrame contains expected columns
    expected_columns = {'title', 'description', 'published_at', 'sentiment_score', 'sentiment_category', 'url'}
    assert set(df.columns).issuperset(expected_columns)

    # Check if the DataFrame is not empty
    assert not df.empty



# +
import pytest
import pandas as pd

def test_classify_average_sentiment():
    # Create a sample DataFrame with predefined sentiment scores
    data = {
        'sentiment_score': [-0.6, -0.2, 0.0, 0.3, 0.6]
    }
    df = pd.DataFrame(data)

    # Call the function
    average_score, sentiment_level = classify_average_sentiment(df)

    # Check the average score and sentiment level
    assert average_score == pytest.approx(0.02, 0.001)
    assert sentiment_level == "Neutral"

def test_classify_average_sentiment_empty_df():
    # Test with an empty DataFrame
    df_empty = pd.DataFrame()
    average_score, sentiment_level = classify_average_sentiment(df_empty)

    # Check the return values for an empty DataFrame
    assert average_score is None
    assert sentiment_level == "No Data"



# +
import pytest
import pandas as pd

def test_fetch_stock_prices():
    symbol = "AAPL"  # Apple Inc. as a commonly known stock symbol

    # Call the function
    df = fetch_stock_prices(symbol, period='5d')  # Fetch data for the last 5 days

    # Check if the result is a DataFrame
    assert isinstance(df, pd.DataFrame), "The result should be a pandas DataFrame"

    # Check if the DataFrame contains the expected columns
    expected_columns = ['Date', 'Close']
    assert set(df.columns) == set(expected_columns), "DataFrame should have Date and Close columns"

    # Check if the DataFrame is not empty
    assert not df.empty, "The DataFrame should not be empty"

    # Optionally, check the data types of the columns
    assert pd.api.types.is_datetime64_any_dtype(df['Date']), "Date column should be datetime type"
    assert pd.api.types.is_float_dtype(df['Close']), "Close column should be float type"

# -

# !poetry run pytest




src/marketmood_api_client2283/marketmood_api_client2283.py
import pandas as pd
import os
import requests
import json
from textblob import TextBlob
from datetime import datetime
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np


def _fetch_financial_sentiment_data(company_name):
    """
    Fetches financial market data for a specific company and analyzes sentiment of the articles using VADER.
    Retrieves the last 100 news items about the company, including the news time and URL.
    :param company_name: Name of the company.
    :return: DataFrame with article details, sentiment scores, publication times, and URLs.
    """
    api_endpoint = 'https://newsapi.org/v2/everything'
    api_key = "f1d6a48ce4574b4ba30eac6e43162957"  
    params = {
        'q': company_name,
        'apiKey': api_key,
        'language': 'en',
        'sortBy': 'publishedAt',
        'pageSize': 100  # Fetch up to 100 articles
    }

    try:
        response = requests.get(api_endpoint, params=params)
        response.raise_for_status()
    except requests.exceptions.HTTPError as http_err:
        print(f'HTTP error occurred: {http_err}')
        return pd.DataFrame()
    except Exception as err:
        print(f'Other error occurred: {err}')
        return pd.DataFrame()

    articles = response.json().get('articles', [])
    data = []
    analyzer = SentimentIntensityAnalyzer()

    for article in articles:
        title = article.get('title', '') or ''
        description = article.get('description', '') or ''
        published_at = datetime.strptime(article.get('publishedAt', '')[:10], '%Y-%m-%d').date() if article.get('publishedAt') else None
        url = article.get('url', '')

        text = f"{title} {description}".strip()
        sentiment_score = analyzer.polarity_scores(text)['compound']
        sentiment_category = 'Positive' if sentiment_score > 0.05 else 'Negative' if sentiment_score < -0.05 else 'Neutral'

        data.append({
            'title': title,
            'description': description,
            'published_at': published_at,
            'sentiment_score': sentiment_score,
            'sentiment_category': sentiment_category,
            'url': url
        })

    df = pd.DataFrame(data)
    return df


# Example usage
company_name = "Tesla"
df = _fetch_financial_sentiment_data(company_name)
df.head()


def classify_average_sentiment(df):
    """
    Computes the average sentiment score and classifies it into levels.
    :param df: DataFrame with sentiment scores.
    :return: A tuple containing the average sentiment score and its classification.
    """
    if df.empty:
        return (None, "No Data")

    average_score = df['sentiment_score'].mean()

    if average_score <= -0.5:
        sentiment_level = "Strongly Negative"
    elif -0.5 < average_score <= -0.1:
        sentiment_level = "Negative"
    elif -0.1 < average_score < 0.1:
        sentiment_level = "Neutral"
    elif 0.1 <= average_score < 0.5:
        sentiment_level = "Positive"
    else:
        sentiment_level = "Strongly Positive"

    return (average_score, sentiment_level)


# +
# Fetch sentiment data
company_name = "Apple"
sentiment_data = _fetch_financial_sentiment_data(company_name)

# Classify average sentiment
average_sentiment, sentiment_level = classify_average_sentiment(sentiment_data)

print(f"Average Sentiment Score for {company_name}: {average_sentiment}")
print(f"Sentiment Level: {sentiment_level}")

# +
# Get the company's stock price and see the correlation between the sentiment scores and the stock price

# +
import yfinance as yf

def fetch_stock_prices(symbol, period='1mo'):
    stock = yf.Ticker(symbol)
    df = stock.history(period=period)
    df.reset_index(inplace=True)
    return df[['Date', 'Close']]


# -

def align_data(df_sentiment, df_stock_prices):
    df_sentiment['date'] = pd.to_datetime(df_sentiment['published_at']).dt.date
    df_stock_prices['date'] = pd.to_datetime(df_stock_prices['Date']).dt.date
    df_combined = pd.merge(df_sentiment, df_stock_prices, left_on='date', right_on='date')
    return df_combined


def analyze_correlation(df_combined):
    return df_combined[['sentiment_score', 'Close']].corr()



# +
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

def plot_data(df_combined):
    plt.figure(figsize=(12, 6))

    # Ensure data is in correct format for Matplotlib
    dates = df_combined['date'].values
    sentiment_scores = df_combined['sentiment_score'].values
    stock_prices = df_combined['Close'].values

    # Create a plot with two y-axes
    fig, ax1 = plt.subplots()

    color = 'tab:red'
    ax1.set_xlabel('Date')
    ax1.set_ylabel('Sentiment Score', color=color)
    ax1.plot(dates, sentiment_scores, color=color, marker='o')
    ax1.tick_params(axis='y', labelcolor=color)

    ax2 = ax1.twinx()  # Instantiate a second axes that shares the same x-axis
    color = 'tab:blue'
    ax2.set_ylabel('Stock Price', color=color)
    ax2.plot(dates, stock_prices, color=color, marker='o')
    ax2.tick_params(axis='y', labelcolor=color)

    plt.title('Sentiment Score and Stock Price Over Time')
    fig.tight_layout()  
    plt.show()



# -

# # Example Usage

# +
# Example usage
company_name = "TSLA"  # Tesla
stock_symbol = "TSLA"

# Fetch Data
df_stock_prices = fetch_stock_prices(stock_symbol)
df_sentiment = _fetch_financial_sentiment_data(company_name) 

# Align and Combine Data
df_combined = align_data(df_sentiment, df_stock_prices)

# Analyze Correlation
correlation = analyze_correlation(df_combined)
print(correlation)

# Plot Data 
plot_data(df_combined)

# -






